### ISSUE TITLE
Animation can't be tested with GWTTestCase

### ISSUE THREAD

**Issue by aepshteyn** [gwtproject/gwt#9616]
**GWT version:** 2.8.2
**Browser (with version):** HtmlUnit (2.19)
**Operating System:** Windows 8.1

---

##### Description

My unit unit tests involving animations started unexpectedly failing after I upgraded to GWT 2.8.2.  I believe this is because GWT 2.8 uses HtmlUnit 2.19.

I have confirmed that the same tests pass with HtmlUnit 2.15 (by placing the `htmlunit-2.15` jars ahead of the GWT jars on the classpath and providing a modified `RunStyle` implementation).

Here is a very basic [example gist](https://gist.github.com/aepshteyn/3221ed13252bd638c2d5ccf21f890815) that I re-created using a bare-bones GWT project.

The animation clearly doesn't work. Its `onUpdate()` method gets called only once (with the argument `0`), and its `onComplete()` method never gets called.

##### Known workarounds

Use HtmlUnit 2.15 (by placing the `htmlunit-2.15` jars ahead of the GWT jars on the classpath and providing a modified `RunStyle` implementation).

**Comment by tbroyer** [comment]
See https://github.com/gwtproject/gwt/commit/e29062339d6369b58270ac3bf746a590a278787b (and more specifically https://github.com/gwtproject/gwt/commit/e29062339d6369b58270ac3bf746a590a278787b#diff-e42cc02ccb36c576883b9844914466a9)

Actually, I think as a workaround you could also add a `<replace-with>` rebinding rule in your test `gwt.xml` (create one if needed) to replace `AnimationScheduler.AnimationSupportDetector` with one that returns `false` so the scheduler uses the timer-based implementation rather than `requestAnimationFrame`.

Though if your test is about testing the animation itself, you'd maybe rather use a `StubAnimationScheduler` instead.

**Comment by aepshteyn** [comment]
Thanks @tbroyer :+1: 

Your suggestion to force `AnimationSchedulerImplTimer` by replacing `AnimationScheduler.AnimationSupportDetector` worked perfectly for me.  Here is a [gist of that solution](https://gist.github.com/aepshteyn/3caae160b0d647bfef02a0fff5890580) for anyone else who encounters this issue.

I actually didn't know about the existence of any of these options before, and it's good stuff to know; I really appreciate your comment! Would it make sense to document them somewhere in the [Developer's Guide](http://www.gwtproject.org/doc/latest/DevGuide.html)?

 `@DoNotRunWith(Platform.HtmlUnitBug)`  is an interesting take on this problem too :laughing:

**Comment by tbroyer** [comment]
Fwiw, `@DoNotRunWith` **is** documented: http://www.gwtproject.org/doc/latest/DevGuideTestingHtmlUnit.html (the last section is a bit outdated though, currently updating it)

Overriding deferred binding rules is brittle though and as a rule of thumb shouldn't be relied on, and thus shouldn't be documented. Most (if not all) deferred binding rules should be considered implementation details of their respective modules and could break at any time (this has happened before; see https://github.com/gwtproject/gwt/commit/1e0586cc683a501cfde5f2e28325f1201090ef46 for example)
